### PR TITLE
Fix handling of conversions between FP and reals

### LINKED ISSUE #7056: Unsat file with declare-fun becomes sat with define-fun
[fp_ar.smt2.txt](https://github.com/cvc5/cvc5/files/7036213/fp_ar.smt2.txt)

**Describe the bug**

On the attached file, cvc5 produces `unsat`, but if line 10 is commented and line 11 is uncommented (this results in replacing a `declare-fun` by a `define-fun`), it produces `sat`. This looks like a bug to me. Do you agree?

FWIW I believe the two versions of the file should be `unsat`.

**Command line arguments**: `cvc5 fp_ar.smt2`
**cvc5 version/commit**: 2c35a4fd262a7aa1cc3b953265d327de4eae9f8e
**Operating system**: OpenSUSE Leap 15.3

**`configure.sh` options**  
None

**`configure.sh` output**  
```
$ ./configure.sh 
-- The C compiler identification is GNU 7.5.0
-- The CXX compiler identification is GNU 7.5.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc - works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ - works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Building Production build
-- Performing Test HAVE_FLAG_O3
-- Performing Test HAVE_FLAG_O3 - Success
-- Configuring with C flag '-O3'
-- Configuring with CXX flag '-O3'
-- Performing Test HAVE_FLAG_Wall
-- Performing Test HAVE_FLAG_Wall - Success
-- Configuring with C flag '-Wall'
-- Configuring with CXX flag '-Wall'
-- Performing Test HAVE_FLAG_Wno_unused_private_field
-- Performing Test HAVE_FLAG_Wno_unused_private_field - Success
-- Configuring with C flag '-Wno-unused-private-field'
-- Configuring with CXX flag '-Wno-unused-private-field'
-- Performing Test HAVE_FLAG_fexceptions
-- Performing Test HAVE_FLAG_fexceptions - Success
-- Configuring with C flag '-fexceptions'
-- Performing Test HAVE_FLAG_Wsuggest_override
-- Performing Test HAVE_FLAG_Wsuggest_override - Success
-- Configuring with CXX flag '-Wsuggest-override'
-- Performing Test HAVE_FLAG_Wnon_virtual_dtor
-- Performing Test HAVE_FLAG_Wnon_virtual_dtor - Success
-- Configuring with CXX flag '-Wnon-virtual-dtor'
-- Performing Test HAVE_FLAG_Wimplicit_fallthrough
-- Performing Test HAVE_FLAG_Wimplicit_fallthrough - Success
-- Configuring with C flag '-Wimplicit-fallthrough'
-- Configuring with CXX flag '-Wimplicit-fallthrough'
-- Performing Test HAVE_FLAG_Wshadow
-- Performing Test HAVE_FLAG_Wshadow - Success
-- Configuring with C flag '-Wshadow'
-- Configuring with CXX flag '-Wshadow'
-- Performing Test HAVE_FLAG_fno_extern_tls_init
-- Performing Test HAVE_FLAG_fno_extern_tls_init - Success
-- Configuring with CXX flag '-fno-extern-tls-init'
-- Performing Test HAVE_FLAG_Wno_class_memaccess
-- Performing Test HAVE_FLAG_Wno_class_memaccess - Success
-- Configuring with CXX flag '-Wno-class-memaccess'
-- Disabling unit tests since assertions are disabled.
-- Found PythonInterp: /home/kanig/install/e3/bin/python3 (found suitable version "3.8.3", minimum required is "3") 
-- Found GMP 6.1.2: /usr/lib64/libgmp.so
-- Looking for getc_unlocked
-- Looking for getc_unlocked - not found
-- Building CaDiCaL rel-1.4.1: /home/kanig/others/cvc5/build/deps/lib/libcadical.a
-- Building Poly f543721215ec17a724dc86820a0430233931a637: /home/kanig/others/cvc5/build/deps/lib/libpicpoly.a
-- Building SymFPU: /home/kanig/others/cvc5/build/deps/include/
-- Performing Test CVC5_NEED_INT64_T_OVERLOADS
-- Performing Test CVC5_NEED_INT64_T_OVERLOADS - Failed
-- Performing Test CVC5_NEED_HASH_UINT64_T_OVERLOAD
-- Performing Test CVC5_NEED_HASH_UINT64_T_OVERLOAD - Failed
-- Looking for unistd.h
-- Looking for unistd.h - found
-- Looking for C++ include ext/stdio_filebuf.h
-- Looking for C++ include ext/stdio_filebuf.h - found
-- Looking for clock_gettime
-- Looking for clock_gettime - found
-- Looking for ffs
-- Looking for ffs - found
-- Looking for optreset
-- Looking for optreset - not found
-- Looking for sigaltstack
-- Looking for sigaltstack - found
-- Looking for strerror_r
-- Looking for strerror_r - found
-- Looking for strtok_r
-- Looking for strtok_r - found
-- Looking for setitimer
-- Looking for setitimer - found
-- Performing Test STRERROR_R_CHAR_P
-- Performing Test STRERROR_R_CHAR_P - Failed
-- Found Git: /usr/bin/git (found version "2.31.1") 
-- Found Java: /usr/lib64/jvm/jre-11-openjdk/bin/java (found version "11.0.11") found components: Runtime 
-- Building ANTLR3 runtime: /home/kanig/others/cvc5/build/deps/lib64/libantlr3c.a
-- Downloading ANTLR3 JAR: /home/kanig/others/cvc5/build/deps/share/java/antlr-3.4-complete.jar
-- Performing Test COMPILER_HAS_HIDDEN_VISIBILITY
-- Performing Test COMPILER_HAS_HIDDEN_VISIBILITY - Success
-- Performing Test COMPILER_HAS_HIDDEN_INLINE_VISIBILITY
-- Performing Test COMPILER_HAS_HIDDEN_INLINE_VISIBILITY - Success
-- Performing Test COMPILER_HAS_DEPRECATED_ATTR
-- Performing Test COMPILER_HAS_DEPRECATED_ATTR - Success

cvc5 1.0-prerelease

Build profile             : production

GPL                       : off
Best configuration        : off

Assertions                : off
Debug symbols             : off
Debug context mem mgr     : off

Dumping                   : on
Muzzle                    : off
Statistics                : on
Tracing                   : off

ASan                      : off
UBSan                     : off
TSan                      : off
Coverage (gcov)           : off
Profiling (gprof)         : off
Unit tests                : off
Valgrind                  : off

Shared libs               : on
Static binary             : off
Python bindings           : off
Java bindings             : off
Python2                   : off

ABC                       : off
CryptoMiniSat             : off
GLPK                      : off
Kissat                    : off
LibPoly                   : on (local)
CoCoALib                  : off

Build libcvc5 only        : off
MP library                : gmp (system)
Editline                  : off

Api docs                  : off


CPPLAGS (-D...): NDEBUG CVC5_DUMPING CVC5_STATISTICS_ON CVC5_USE_POLY
CXXFLAGS       : -O3 -Wall -Wno-unused-private-field -Wsuggest-override -Wnon-virtual-dtor -Wimplicit-fallthrough -Wshadow -fno-extern-tls-init -Wno-class-memaccess
CFLAGS         : -O3 -Wall -Wno-unused-private-field -fexceptions -Wimplicit-fallthrough -Wshadow
Linker flags   : 

Install prefix : /usr/local

cvc5 license: modified BSD

Note that this configuration is NOT built against any GPL'ed libraries, so
it is covered by the (modified) BSD license.  This is, however, not the best
performing configuration of cvc5.  To build against GPL'ed libraries which
improve cvc5's performance, re-configure with '-DENABLE_GPL -DENABLE_BEST'.

Now change to 'build' and type 'make', followed by 'make check' or 'make install'.

-- Configuring done
-- Generating done
-- Build files have been written to: /home/kanig/others/cvc5/build
```

### PR DESCRIPTION
**Pull request by 4tXJ7f**
Fixes #7056. Currently, we introduce a UF for conversions between FP
numbers and reals. This is done as part of `ppRewrite()`. The problem is
that terms sent to `ppRewrite()` are not fully preprocessed yet and the
FP theory solver was storing terms that were not fully preprocessed in a
map for later lookup. For the issue at hand, the conversion term
contained an `ite`, which was later removed. As a result, the theory
solver did not consider the term to be relevant when refining
abstractions. This commit changes the handling of conversion functions
to instead adding purification lemmas for conversion terms when they are
registered. The purification lemmas are needed, because otherwise, we
can't retrieve the model values for the term without the rewriter
interferring.

### CODE REVIEW

**Review by ajreynol — CHANGES_REQUESTED** [review]

File: src/theory/fp/theory_fp.cpp [hunk 1]
```diff
@@ -587,80 +472,133 @@ void TheoryFp::registerTerm(TNode node)
...
+    return;
+  }
+
+  Kind k = node.getKind();
+  Assert(k != kind::FLOATINGPOINT_TO_FP_GENERIC && k != kind::FLOATINGPOINT_SUB
+         && k != kind::FLOATINGPOINT_EQ && k != kind::FLOATINGPOINT_GEQ
+         && k != kind::FLOATINGPOINT_GT);
+
+  bool success = d_registeredTerms.insert(node);
+  (void)success;  // Only used for assertion
```
[the remark this answers is not in the document]
Comment (4tXJ7f) [reply src/theory/fp/theory_fp.cpp:486]: This is just moved from existing code but I'll update.

File: src/theory/fp/theory_fp.cpp [hunk 2]
```diff
@@ -587,80 +472,133 @@ void TheoryFp::registerTerm(TNode node)
...
+    handleLemma(nm->mkNode(kind::EQUAL, node, equalityAlias),
+                InferenceId::FP_REGISTER_TERM);
+  }
+  else if (k == kind::FLOATINGPOINT_TO_REAL_TOTAL)
+  {
+    // Purify (fp.to_real x)
+    NodeManager* nm = NodeManager::currentNM();
+    SkolemManager* sm = nm->getSkolemManager();
+    Node sk = sm->mkPurifySkolem(node, "to_real", "fp purify skolem");
+    handleLemma(node.eqNode(sk), InferenceId::FP_REGISTER_TERM);
```
[the remark this answers is not in the document]
Comment (4tXJ7f) [reply src/theory/fp/theory_fp.cpp:549]: This code is part of `registerTerm()`, which is called in `preRegisterTerm()` and `preNotifyFact()`, so I think that sending a lemma is correct, no?

[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/fp/theory_fp.cpp:548]: Ah, in fact, you've moved it in this PR, right?  In this case, there is no issue.

**Review by aniemetz — COMMENTED** [review]
I think the comment on line 639 in `handleLemma` is obsolete with these changes.

**Review by martin-cs — APPROVED** [review]
Thanks to @4tXJ7f this one seems like it was a real challenge to find!  I /think/ I understand the fix; rather than creating new uninterpreted functions it uses Purify Skolem.  Otherwise the (static) lemmas it generates and the refinement are unchanged.

**Review by ajreynol — APPROVED** [review]

**Pull request merged**